### ISSUE TITLE
Is this library React compatible?

### ISSUE THREAD

**Issue by ryami333** [developit/unistore#4]
As I understand it, Preact is meant to be approximately API compatible with React, so why is it the specific target of this library? Is there any reason why it shouldn't work with React anyway?

**Comment by developit** [comment]
Three main reasons right now:

1. It imports preact. When you're building a library, you sadly have to pick one or the other since there's no way to import an abstract implementation.  This could by solved through aliasing `preact` to `react.
2. It might use `Component` in a way React wouldn't allow (incomplete inheritance) as a way to save bytes.
3. `<Provider>` doesn't have `contextTypes`, which are required in React.

`#1` and `#3` could be solved by adding a second entry to the module (`unistore/react`), but I'm not sure how that would work without either duplicating the code or increasing filesize unnecessarily.

**Comment by kingdaro** [comment]
It feels like the best way would be to have two packages, like redux and friends do. How much of a development overhead would that produce?

**Comment by hex13** [comment]
> It imports preact. When you're building a library, you sadly have to pick one or the other since 
> there's no way to import an abstract implementation. 

Why not to use dependency injection design pattern and just inject preact/react/whatever during the construction?

```javascript
const unistore = require('unistore');
const { Provider, createStore, connect } = unistore(preact);
```

**Comment by drcmda** [comment]
I would also favour if Provider and connect are outside of the libs scope. Then it doesn't matter how react or preact do things. Reduxes greatest strength is that it is independent, and that in turn has inspired and kickstarted an eco system around it. A state manager around a single lib is a liability i wouldn't risk. I really enjoy how focussed it is otherwise. I will definitively watch the progress of it and hope it makes the right decisions.

**Comment by developit** [comment]
I'm up for splitting `createStore` into its own lib and having entrypoints for the library integrations.

I don't want to do DI since it'll bloat the size considerably.  Also as I mentioned above, the React implementation will require PropTypes, which are dead weight in Preact.

**Comment by cfenzo** [comment]
Here is a take on separation of core and bindings..
Preact: https://codesandbox.io/s/4z712kp769
React: https://codesandbox.io/s/545829n03p

(couldn't find a good way to showcase both in the same codesandbox)

**Comment by matheusml** [comment]
The last comment by @developit was exactly what we have done in [Redux Zero](https://github.com/concretesolutions/redux-zero).

I think this is the best approach for two reasons:
1. People import whatever they with no extra weight
2. The API can be different. For example, in Preact/React we can have a `Provider and a connect` function. But for Vue.js we only have a `connect` function an so on.

**Comment by developit** [comment]
Fixed and released in [3.0.0](https://github.com/developit/unistore/releases/tag/3.0.0).
